subxt/src/tx/signer.rs

```diff
@@ -29,55 +24,82 @@ pub trait Signer<T: Config> {
     fn sign(&self, signer_payload: &[u8]) -> T::Signature;
 }
 
-/// A [`Signer`] implementation that can be constructed from an [`Pair`].
-#[derive(Clone, Debug)]
-pub struct PairSigner<T: Config, P: Pair> {
-    account_id: T::AccountId,
-    signer: P,
-}
+#[cfg(feature = "substrate-compat")]
+pub use pair_signer::PairSigner;
 
-impl<T, P> PairSigner<T, P>
-where
-    T: Config,
-    T::Signature: From<P::Signature>,
-    <T::Signature as Verify>::Signer:
-        From<P::Public> + IdentifyAccount<AccountId = T::AccountId>,
-    P: Pair,
-{
-    /// Creates a new [`Signer`] from a [`Pair`].
-    pub fn new(signer: P) -> Self {
-        let account_id =
-            <T::Signature as Verify>::Signer::from(signer.public()).into_account();
-        Self { account_id, signer }
-    }
+// A signer suitable for substrate based chains. This provides compatibility with Substrate
+// packages like sp_keyring and such, and so relies on sp_core and sp_runtime to be included.
+#[cfg(feature = "substrate-compat")]
+mod pair_signer {
+    use super::Signer;
+    use crate::Config;
+    use sp_core::Pair as PairT;
+    use sp_runtime::{
+        traits::{
+            IdentifyAccount,
+            Verify,
+        },
+        AccountId32 as SpAccountId32,
+        MultiSignature as SpMultiSignature,
+    };
 
-    /// Returns the [`Pair`] implementation used to construct this.
-    pub fn signer(&self) -> &P {
-        &self.signer
+    /// A [`Signer`] implementation that can be constructed from an [`sp_core::Pair`].
+    #[derive(Clone, Debug)]
+    pub struct PairSigner<T: Config, Pair> {
+        account_id: T::AccountId,
+        signer: Pair,
     }
 
-    /// Return the account ID.
-    pub fn account_id(&self) -> &T::AccountId {
-        &self.account_id
-    }
-}
+    impl<T, Pair> PairSigner<T, Pair>
+    where
+        T: Config,
+        T::AccountId: From<SpAccountId32>,
```

Comment:
You create a PairSigner by passing a Pair implementation whose signature type can provide an `sp_runtime::MultiSignature`, where that has a Signer type that can provide back an `sp_runtime::AccountId32`. You must be able to convert `sp_runtime::AccountId32` into our configured account type, `T::AccountId`.

Below you can see that you also need to be able to convert a `Pair::Signature`into an `sp_runtime::MultiSignature`, and take that into our configured `T::Signature` type.

In other words, we use the substrate types as a go-between. I'm hoping that this won't limit the use cases for this type, but I'm not sure offhand so would love to hear any feedback or thoughts here!

Reply:
I'll see how hard it is to impl the substrate signer stuff on MultiSignature etc and then maybe we can avoid any intermediate types.

Reply:
I ended up simplifying them a bit and adding a couple of extra `From` impls to hopefully make the different signature types in Substrate "just work".